text/0000-sync-networked-vars.md

```diff
@@ -0,0 +1,110 @@
+
+This sets up MLAPI for upcoming features like snapshot.
+
+# Dependency
+[dependency]: #dependency
+
+This feature depends on a Tick System that would provide:
+- A discrete measurement of the local time, based on an integer number of tick elapsed
+- The tick number being updated at the very beginning of each frame and stable throughout the frame
+- A tick that is based on the real time elapsed since a fixed point in the past 
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Some definitions:
+
+#### Local tick
+The tick number at which a variable was modified, relative to the tick system on this machine. If I get an update for a variable I changed, this tell me: when did I change it?
+
+#### Remote tick
+The tick number at which a variable was relayed, relative to the tick system on the remote machine. If I get an update from the server, this tells me the local tick on the server when it sent the update.
+
+#### Network Tick wrapping 
+
+The tick value sent will be a rolling value, probably 8- or 16-bit. This will be hidden internally, but will keep the bandwidth usage in check.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+We suggest changing:
+
+- in `Runtime/NetworkedVar/INetworkedVar.cs`, `INetworkedVar` `ReadField()` to become
+
+      public void ReadField(Stream stream, ushort localTick, ushort remoteTick);
+
+- in `Runtime/NetworkedVar/INetworkedVar.cs`, `INetworkedVar` `ReadDelta()` to become
+
+      public void ReadDelta(Stream stream, bool keepDirtyDelta, ushort localTick, ushort remoteTick);
+
+- `NetworkedVar` to include two new properties:
+
+      public ushort LocalTick { get; internal set; }
+      public ushort RemoteTick { get; internal set; }
+      
+ - `SendTickrate` and `LastSyncedTime` to be removed from `NetworkedVar`
+
+Below diagram shows a couple NetworkedVariable updates with example tick numbers.
+
+![](0000-sync-networked-vars/netvars.png)
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This increases slightly the bandwidth usage of _individual_ NetworkedVariable update it also renders the NetworkedVariables system dependent on the tick system.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+This proposal makes the NetworkedVariables a bit more complex. But it sets a foundation over which other features can be built without causing the NetworkedVariables to be modified further. Snapshotting, client-side interpolation, lag compensation, replay and rollback are examples features that will leverage this. 
+
+# Prior art
+[prior-art]: #prior-art
+
+This is a pretty typical setup seen in other proprietary network engines.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+Do we require the tick to advance at the same rate on the client and server machine. Can we support an asymmetry in tick rate between client and server? Is this important?
```

Comment:
Yes: client input sendrate vs server snapshot sendrate.
Yes: this is important :)

Reply:
I need help becoming a believer in this :).  Seems like it will make even things like debugging harder if we're looking at server tick 123 and having to correlate it to client tick 456

Reply:
The way I see it, we sync the netvars as a firs task/RFC, then once we have the tick system, we consider aligning the ticks when a game is joined. I basically tried taking the path of least resistance: make something that could work if the tick were offset'ed. If we decide to align them later, fine too.